substrate/runtime/contract/src/vm.rs

```diff
@@ -0,0 +1,785 @@
+	///
+	/// The contract can designate some buffer at the execution time via a special function.
+	/// If contract called this function with non-empty buffer it will be copied here.
+	///
+	/// Note that gas is already charged for returning the data.
+	pub fn return_data(&self) -> &[u8] {
+		&self.return_data
+	}
+}
+
+/// Execute the given code as a contract.
+pub fn execute<'a, T: Trait, E: Ext<T>>(
+	code: &[u8],
+	ext: &'a mut E,
+	gas_meter: &mut GasMeter<T>,
+) -> Result<ExecutionResult, Error> {
+	// TODO: Receive data as an argument
+
+	// ext_gas(amount: u32)
+	//
+	// Account for used gas. Traps if gas used is greater than gas limit.
+	//
+	// - amount: How much gas is used.
+	fn ext_gas<T: Trait, E: Ext<T>>(
+		e: &mut Runtime<T, E>,
+		args: &[sandbox::TypedValue],
+	) -> Result<sandbox::ReturnValue, sandbox::HostError> {
+		let amount = args[0].as_i32().unwrap() as u32;
+		let amount = <T::Gas as As<u32>>::sa(amount);
+
+		match e.gas_meter.charge(amount) {
+			GasMeterResult::Proceed => Ok(sandbox::ReturnValue::Unit),
+			GasMeterResult::OutOfGas => Err(sandbox::HostError),
+		}
+	}
+
+	// ext_put_storage(location_ptr: u32, value_non_null: u32, value_ptr: u32);
+	//
+	// Change the value at the given location in storage or remove it.
+	//
+	// - location_ptr: pointer into the linear
+	//   memory where the location of the requested value is placed.
+	// - value_non_null: if set to 0, then the entry
+	//   at the given location will be removed.
+	// - value_ptr: pointer into the linear memory
+	//   where the value to set is placed. If `value_non_null` is set to 0, then this parameter is ignored.
+	fn ext_set_storage<T: Trait, E: Ext<T>>(
+		e: &mut Runtime<T, E>,
+		args: &[sandbox::TypedValue],
+	) -> Result<sandbox::ReturnValue, sandbox::HostError> {
+		let location_ptr = args[0].as_i32().unwrap() as u32;
+		let value_non_null = args[1].as_i32().unwrap() as u32;
+		let value_ptr = args[2].as_i32().unwrap() as u32;
+
+		let mut location = [0; 32];
+
+		e.memory().get(location_ptr, &mut location)?;
+
+		let value = if value_non_null != 0 {
+			let mut value = [0; 32];
+			e.memory().get(value_ptr, &mut value)?;
+			Some(value.to_vec())
+		} else {
+			None
+		};
+		e.ext.set_storage(&location, value);
+
+		Ok(sandbox::ReturnValue::Unit)
+	}
+
+	// ext_get_storage(location_ptr: u32, dest_ptr: u32);
+	//
+	// Retrieve the value at the given location from the strorage.
+	// If there is no entry at the given location then all-zero-value
+	// will be returned.
+	//
+	// - location_ptr: pointer into the linear
+	//   memory where the location of the requested value is placed.
+	// - dest_ptr: pointer where contents of the specified storage location
+	//   should be placed.
+	fn ext_get_storage<T: Trait, E: Ext<T>>(
+		e: &mut Runtime<T, E>,
+		args: &[sandbox::TypedValue],
+	) -> Result<sandbox::ReturnValue, sandbox::HostError> {
+		let location_ptr = args[0].as_i32().unwrap() as u32;
```

Comment:
same same.

Reply:
let's move this discussion in [this thread](https://github.com/paritytech/polkadot/pull/345#discussion_r203060428).